feat(php): Set PHP version per site

Mte90 opened the pull request:
Ref: https://github.com/Varying-Vagrant-Vagrants/VVV/issues/2367
Part of: https://github.com/Varying-Vagrant-Vagrants/varyingvagrantvagrants.org/pull/207

## Checks

<!--  Have you: -->
* [x] I've updated the changelog.
* [x] I've tested this PR
* [x] This PR is for the `develop` branch not the `stable` branch.
* [x] This PR is complete and ready for review.

Mte90 commented:
It isn't yet ready, after the first on set the php version I get:
```
update-alternatives: error: alternative /usr/bin/php for php not registered; not setting
```

On it

Mte90 commented:
Now seems working, now check also if the same version is already configured and avoid to set it again at the same version

tomjn commented on config/default-config.yml:
```diff
@@ -36,6 +36,8 @@ sites:
     description: "A standard WP install, useful for building plugins, testing things, etc"
     repo: https://github.com/Varying-Vagrant-Vagrants/custom-site-template.git
+    php: 7.4 # change the PHP version to use for the provision
+    nginx_upstream: 7.4 # change the php version to use for nginx
```
the hope is we can remove the nginx upstream because PHP was set to 7.4, as a result the provisioner can infer what the upstream value should be

Mte90 replied:
Right now they are 2 different values but I am not sure what this feedback is just a note or something to do

tomjn replied:
I don't think we should mention the nginx upstream here, and it should work without it and just `php: 7.4`

tomjn replied:
So in the future the docs would say just put `php: 8.0` and it would run the provisioners in 8.0 and the nginx config would also use PHP 8 instead of needing an nginx upstream value.

Then in a future release we can auto-run the PHP 8 extension before the sites provision based on what the config says

Mte90 replied:
So with a single value change both the parameters.
Maybe it's the case to open a ticket just for this.

tomjn replied:
but we did open a ticket for this, https://github.com/Varying-Vagrant-Vagrants/VVV/issues/2367 which is what this PR is meant to implement

tomjn replied:
This is still a hard blocker and a major issue, it implies a misunderstanding of what #2367 was asking for. The whole point of the issue is that you no longer need to specify an Nginx upstream to change your PHP version, you only specify an upstream if you specifically and only want to change the upstream, not because you want to change the PHP version used.

Mte90 replied:
The problem with this is: what if I want that the provision is executed with a php version and nginx with another one?

tomjn replied:
then it would use the nginx upstream in the config because the user has explicitly chosen the value.

 - `php` tells VVV which version of PHP to use for the site
 - `nginx_upstream` allows you to override a value in the nginx config

Right now we use `nginx_upstream` to set the version of PHP used on frontend requests because no other options exist, it's a workaround. This issue is meant to fix that by providing a `php:` parameter

tomjn commented on provision/provision-site.sh:
```diff
@@ -30,6 +30,26 @@ VVV_CONFIG=/vagrant/config.yml
+  local value=$(shyaml -q get-value "sites.${SITE_ESCAPED}.${1}" "${2}" < ${VVV_CONFIG})
+  echo "${value}"
+}
+php_version=$(readlink -f /usr/bin/php)
```
It's possible that a provisioner somewhere else either failed or interfered with this, so we should always change the PHP version used if it's not the default version, e.g. if a previous site switched the PHP version to 8.0 then failed it would test against 8.0 rather than the default of 7.4

Mte90 replied:
the next line does this, check if the php version is different from the default otherwise reset it

tomjn commented on provision/provision-site.sh:
```diff
@@ -30,6 +30,26 @@ VVV_CONFIG=/vagrant/config.yml
+if [[ $php_version != *"${VVV_BASE_PHPVERSION}"* ]]; then
+  DEFAULTPHP=$(vvv_get_site_config_value 'php' "${VVV_BASE_PHPVERSION}")
+  echo " * Setting the default PHP CLI version ( ${DEFAULTPHP} ) for this site"
+  noroot update-alternatives --set php "/usr/bin/php${DEFAULTPHP}"
+  noroot update-alternatives --set phar "/usr/bin/phar${DEFAULTPHP}"
+  noroot update-alternatives --set phar.phar "/usr/bin/phar.phar${DEFAULTPHP}"
+  noroot update-alternatives --set phpize "/usr/bin/phpize${DEFAULTPHP}"
+  noroot update-alternatives --set php-config "/usr/bin/php-config${DEFAULTPHP}"
```
does `noroot` make sense here? We run these commands a lot, maybe it's time to turn them into functions, perhaps:

`vvv_set_php_cli_version` and `vvv_reset_php_cli_version`, where the following behaviour happens:

```sh
vvv_set_php_cli_version "7.2" # php is now at 7.2
vvv_reset_php_cli_version # php is now at the default 7.4
vvv_set_php_cli_version  # no parameter so it calls vvv_reset_php_cli_version
```

Mte90 replied:
I preferred to keep the function as you suggest but read the parameter by the config instead as parameter

tomjn replied:
the concern is more that broken provisioning might break the expectations here causing more broken provisioning which just sounds painful to debug :(

tomjn commented on provision/provision-site.sh:
```diff
@@ -504,4 +516,8 @@ if [ "${SUCCESS}" -ne "0" ]; then
 fi
 
+if [[ $php_version != *"${VVV_BASE_PHPVERSION}"* ]]; then
+  noroot /srv/config/homebin/vvv_restore_php_default
```
unsure about the `noroot`, but it looks like one of the commands from my previous feedback! Great minds think alike :)

tomjn replied:
Maybe this check should be inside `vvv_restore_php_default` though?

tomjn reviewed (changes requested):
some notes

tomjn commented on provision/provision-site.sh:
```diff
@@ -30,6 +30,29 @@ VVV_CONFIG=/vagrant/config.yml
+}
+
+# @description Reset the PHP global version to the default
+function vvv_set_php_cli_version() {
```
This function is not indented correctly

tomjn reviewed (changes requested):
I'm seeing no adjustment of the nginx upstream here, this is all I should need to specify that a site uses PHP 8.0:

```yaml
  phpeight:
    repo: https://github.com/Varying-Vagrant-Vagrants/custom-site-template.git
    php: 8.0
    hosts:
      - phpeight.test
```
But, this PR only changes the PHP version the provisioner uses, and does not touch the Nginx upstream.

 - If I create a site, the nginx upstream will use the default PHP of VVV
 - if I create a site with `php: 8.0` the nginx upstream will use PHP 8.0
 - if i create a site with an `nginx_upstream: abc` it will use `abc` even if `php: 8.0` is set

```
nginx upstream = default PHP
if php is set then
    nginx upstream = php{php config value}
if nginx upstream is set then
   nginx upstream = {nginx upstream config value}
```

Mte90 commented:
Fixed the linting and removed the example for nginx_upstream parameter in default-config.yml.
Also if nginx_upstream is not set and php parameter is set use that value for nginx_upstream (I have to test it).

tomjn commented on provision/provision-site.sh:
```diff
@@ -102,6 +133,10 @@ function vvv_provision_site_nginx_config() {
   if [ 'php' != "${NGINX_UPSTREAM}" ] && [ ! -f "/etc/nginx/upstreams/${NGINX_UPSTREAM}.conf" ]; then
     vvv_error " * Upstream value '${NGINX_UPSTREAM}' doesn't match a valid upstream. Defaulting to 'php'.${CRESET}"
     NGINX_UPSTREAM='php'
+    DEFAULTPHP=$(vvv_get_site_config_value 'php' "")
+    if [ '' != "${DEFAULTPHP}" ]; then
+      NGINX_UPSTREAM=$DEFAULTPHP
+    fi
   fi
   sed -i "s#{upstream}#${NGINX_UPSTREAM}#"  "${TMPFILE}"
```
I don't think this will work, if no `nginx_upstream` is provided then its value will default to `php` and none of this code will run. This code is checking if the nginx upstream the user has specified in the config actually exists, and resets it to `php` if it does not

tomjn commented on provision/provision-site.sh:
```diff
@@ -102,6 +133,10 @@ function vvv_provision_site_nginx_config() {
   if [ 'php' != "${NGINX_UPSTREAM}" ] && [ ! -f "/etc/nginx/upstreams/${NGINX_UPSTREAM}.conf" ]; then
```
```suggestion
  # if nginx upstream is not 'php' then it has been changed in config.yml
  if [ 'php' != "${NGINX_UPSTREAM}" ] && [ ! -f "/etc/nginx/upstreams/${NGINX_UPSTREAM}.conf" ]; then
```

tomjn commented on provision/provision-site.sh:
```diff
@@ -102,6 +133,10 @@ function vvv_provision_site_nginx_config() {
+    if [ '' != "${DEFAULTPHP}" ]; then
+      NGINX_UPSTREAM=$DEFAULTPHP
+    fi
   fi
```
```suggestion
  else
    # if php: is configured, set the upstream to match
    if [ '' != "${DEFAULTPHP}" ]; then
      # TODO: Check this version of PHP is actually installed
      NGINX_UPSTREAM="${DEFAULTPHP}"
    fi
  fi
```

tomjn replied:
also, aren't PHP versions of the form `8.0` but nginx upstreams of the form `php80`? We have to do some conversion

Mte90 replied:
you are right, fixed in a commit

tomjn commented on provision/provision-site.sh:
```diff
@@ -484,6 +511,7 @@ function vvv_custom_folders() {
 }
 
 # -------------------------------
+vvv_set_php_cli_version()
```
needs a doc block

tomjn replied:
but no docblock, so it won't have auto-generated documentation when the docs site is updated

tomjn replied:
See https://github.com/Varying-Vagrant-Vagrants/varyingvagrantvagrants.org/blob/main/inject-repo-documentation.sh and https://github.com/reconquest/shdoc

tomjn replied:
and https://varyingvagrantvagrants.org/docs/en-US/function-reference/provision-provision-site/

Mte90 commented:
Fixed the issues you reported, I am not sure if I fixed the doc stuff.

tomjn commented on provision/provision-site.sh:
```diff
@@ -23,13 +23,38 @@ NGINX_UPSTREAM=$6
+  echo "${value}"
+}
+
+# @description Reset the PHP global version to the default
```
not quite true as it checks the site config value, so if the default is 7.4 but my site says `php: 8.0` then this function sets PHP to v8, not v7.4

tomjn replied:
```suggestion
# @description sets a sites PHP version as the global version, or to the VVV default if none is specified
```

tomjn commented on provision/provision-site.sh:
```diff
@@ -99,10 +124,17 @@ function vvv_provision_site_nginx_config() {
   sed -i "s#{vvv_site_name}#${SITE_NAME}#"  "${TMPFILE}"
   sed -i "s#{vvv_hosts}#${VVV_HOSTS}#"  "${TMPFILE}"
 
+  # if nginx upstream is not 'php' then it has been changed in config.yml
```
```suggestion
  # check if the nginx upstream value has been set and is valid
```

tomjn commented on provision/provision-site.sh:
```diff
@@ -99,10 +124,17 @@ function vvv_provision_site_nginx_config() {
     vvv_error " * Upstream value '${NGINX_UPSTREAM}' doesn't match a valid upstream. Defaulting to 'php'.${CRESET}"
     NGINX_UPSTREAM='php'
   fi
+
+  # if php: is configured, set the upstream to match
+  if [ "${DEFAULTPHP}" != "${VVV_DEFAULTPHP}" ]; then
+    NGINX_UPSTREAM="php${DEFAULTPHP}"
+    NGINX_UPSTREAM=$(echo $NGINX_UPSTREAM | tr --delete .)
+  fi
```
this should happen beforehand otherwise I could do the following:

```yaml
    php: 74.0
```
And Nginx would fail to load as there is no PHP v74.0

Mte90 replied:
ops I switched the lines

tomjn reviewed (changes requested):
Some small changes to make it handle failure gracefully if a user puts a silly value in like `v800.0` or `php: banana`

Mte90 commented:
Added a version validation that check if the string has a length different from 3. I am not sure if this way is the best one anyway (require some testing).

tomjn commented:
Things look good, I need to test before I can approve but I think this PR is the priority for releasing 3.10.1

tomjn commented:
Looks like there are some CI issues finding functions in site templates:

```
    default:  * wordpress-one provisioner clone successful
    default: PHP Notice:  Undefined index: HTTP_HOST in /srv/www/wordpress-one/public_html/wp-includes/functions.php on line 6052
    default: Notice: Undefined index: HTTP_HOST in /srv/www/wordpress-one/public_html/wp-includes/functions.php on line 6052
    default:  * sourcing of vvv-init.sh reported success
    default:  * VVV is adding an Nginx config from /srv/www/wordpress-one/provision/vvv-nginx.conf
    default: nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
    default: nginx: configuration file /etc/nginx/nginx.conf test is successful
    default: /tmp/vagrant-shell: line 556: vvv_restore_php_default: command not found
    default:  ! The 'site-wordpress-one' provisioner ran into problems, the full log is available at '/var/log/provisioners/2022.08.01_12-27-27/provisioner-site-wordpress-one.log'. It completed in 17 seconds.
```

Mte90 commented:
I am not able to replicate your issue maybe in the PR the permissions on that file are wrong?

In the meantime I found another bug and fixed it.

tomjn commented:
Local tests failed with:

```
    default:  ✔ The 'site-wordpress-two' provisioner completed in 2 seconds.
==> default: Running provisioner: site-phpeight (shell)...
    default: Running: /var/folders/dw/wdtnsdn154db04ymxv90t9tm0000gn/T/vagrant-shell20220907-23209-1feiyc8.sh
    default:  ▷ Running the 'site-phpeight' provisioner...
    default: /tmp/vagrant-shell: line 51: vvv_warning: command not found
    default:  ! The 'site-phpeight' provisioner ran into problems, the full log is available at '/var/log/provisioners/2022.09.07_13-44-29/provisioner-site-phpeight.log'. It completed in 0 seconds.
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

tomjn commented:
Fixed a `vvv_warning` -> `vvv_warn`, though `8.0` is probably being interpreted as a number as it considers it the wrong format. We can't expect users to always wrap it in quotes so some string manipulation will be needed to format the value in the 0.0 format

tomjn reviewed (changes requested):
Needs to be able to handle numbers in the value such as `8.1` or `8.0` without wrapping in string quotes e.g. `"8.0"`

tomjn commented:
I changed it to:

```
    length=$(echo "$DEFAULTPHP" | wc -c)
    if [[ $length > '3' ]]; then
      vvv_warn " ! Warning: PHP version defined is using a wrong format: '${DEFAULTPHP}' with length '${length}'"
    else
```

and got:

```
    default:  ▷ Running the 'site-phpeight' provisioner...
    default:  ! Warning: PHP version defined is using a wrong format: '8.0' with length '4'
    default:  * Pulling down the master branch of https://github.com/Varying-Vagrant-Vagrants/custom-site-template.git
```

Looks like there's a newline or terminating character that isn't being handled. Incrementing `3` to `4` leads to an unhandled error:

```
    default: Running: /var/folders/dw/wdtnsdn154db04ymxv90t9tm0000gn/T/vagrant-shell20220907-30221-7dsbtn.sh
    default:  ▷ Running the 'site-phpeight' provisioner...
    default:  ! The 'site-phpeight' provisioner ran into problems, the full log is available at '/var/log/provisioners/2022.09.07_13-50-27/provisioner-site-phpeight.log'. It completed in 0 seconds.
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

```

Mte90 commented:
I added the removal of hidden breaklines but I wasn't able to find a way to validate a "version string" expect for the length.

tomjn commented:
@Mte90 would it make more sense to validate the presence of that version of PHP rather than the validity of the version string?

e.g. it doesn't matter if `php : Octopus` is a valid because there'll never be a `phpOctupus` nginx upstream or an `octopus` folder and CLI

Mte90 commented:
Well this simplify a lot but this means that they need to have already provisioned to be able to have that PHP version in VVV.
Anyway is better of my fake solution.

tomjn commented:
@Mte90 extensions always provision before sites so it'll be installed beforehand, but if it is missing it'll just revert to default PHP
